File: odin/accounting/application/use_cases/category_creator.py
import uuid

from nyoibo import Entity, fields

from odin.accounting.domain import Category, CategoryType
from odin.accounts.domain import User
from ..repositories import CategoryRepository


class CategoryCreator(Entity):
    _name = fields.StrField()
    _user = fields.LinkField(to=User)
    _type = fields.StrField(choices=CategoryType)

    def __init__(self, category_repository: CategoryRepository, **kwargs):
        super().__init__(**kwargs)
        self._repository = category_repository

    def create(self) -> Category:
        category = Category(name=self.name, id=uuid.uuid4(), user=self._user, type=self._type)
        self._repository.add(category)
        return category

File: src/accounting/application/use_cases/categorycreator/category_creator.go
package categorycreator

import (
	"context"
	"errors"

	"github.com/google/uuid"
	"raiseexception.dev/odin/src/accounting/application/commands/categorycommand"
	"raiseexception.dev/odin/src/accounting/domain/category"
	"raiseexception.dev/odin/src/accounting/domain/repositories"
)

type CategoryCreator struct {
	command    categorycommand.CategoryCreatorCommand
	repository repositories.CategoryRepository
}

func New(command categorycommand.CategoryCreatorCommand, repository repositories.CategoryRepository) CategoryCreator {
	return CategoryCreator{command: command, repository: repository}
}

func (self CategoryCreator) Create(ctx context.Context) (*categorymodel.Category, error) {
	id, uuidError := uuid.NewV7()
	if uuidError != nil {
		return nil, errors.New("error generating uuid")
	}
	category := categorymodel.New(
		id.String(),
		self.command.Name(),
		self.command.Type(),
		self.command.UserID(),
	)
	err := self.repository.Add(ctx, category)
	if err != nil {
		return nil, err
	}
	return category, nil
}
